sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala

```
@@ -635,150 +644,78 @@ class ResolveSessionCatalog(
-          c.comment, c.external))
-
-      case (None, None) =>
-        Some(buildHiveCatalogTable(
-          ident, schema, c.partitioning, c.bucketSpec, c.properties, SerdeInfo.empty, c.options,
-          c.location, c.comment, c.external))
-
-      case (Some(provider), None) if !isV2Provider(provider) =>
-        Some(buildCatalogTable(
-          ident, schema, c.partitioning, c.bucketSpec, c.properties, provider, c.options,
-          c.location, c.comment, c.external))
-
-      case _ =>
-        None
-    }
-  }
-
-  private def buildCatalogTable(
-      table: TableIdentifier,
-      schema: StructType,
-      partitioning: Seq[Transform],
-      bucketSpec: Option[BucketSpec],
-      properties: Map[String, String],
-      provider: String,
+  private def getStorageFormatAndProvider(
+      provider: Option[String],
       options: Map[String, String],
       location: Option[String],
-      comment: Option[String],
-      external: Boolean): CatalogTable = {
-    // Hive types are allowed if the provider is "hive"
-    if (!DDLUtils.isHiveTable(Some(provider))) {
-      assertNoCharTypeInSchema(schema)
-    }
-
-    if (external) {
-      throw new AnalysisException(s"Operation not allowed: CREATE EXTERNAL TABLE ... USING")
-    }
-
-    val storage = CatalogStorageFormat.empty.copy(
+      maybeSerdeInfo: Option[SerdeInfo],
+      ctas: Boolean): (CatalogStorageFormat, String) = {
+    val nonHiveStorageFormat = CatalogStorageFormat.empty.copy(
+      locationUri = location.map(CatalogUtils.stringToURI),
+      properties = options)
+    val defaultHiveStorage = HiveSerDe.getDefaultStorage(conf).copy(
       locationUri = location.map(CatalogUtils.stringToURI),
       properties = options)
 
-    val tableType = if (location.isDefined) {
-      CatalogTableType.EXTERNAL
+    if (provider.isDefined) {
+      // The parser guarantees that USING and STORED AS/ROW FORMAT won't co-exist.
+      assert(maybeSerdeInfo.isEmpty)
```

This should throw a helpful exception with a real message. Before it was `throw new AnalysisException(s"Cannot create table with both USING $provider and ${serde.describe}")`